### ISSUE TITLE
Config json_encoders does not seem to work on builtin types

### ISSUE THREAD

**Issue by emr-arvig** [pydantic/pydantic#2531]
Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.1
            pydantic compiled: True
                 install path: ../MYPROJECTS/MISC_TESTING/.venv/lib/python3.7/site-packages/pydantic
               python version: 3.7.10 (default, Feb 20 2021, 21:21:24)  [GCC 5.4.0 20160609]
                     platform: Linux-4.4.0-18362-Microsoft-x86_64-with-Ubuntu-16.04-xenial
     optional deps. installed: ['typing-extensions']
```

I am getting data from a database and passing it to Pydantic models which ensure correct output for API calls.

This database stores True/False values as 1/0, so I wanted to input to a model then ensure they were correct JSON for API payloads on output. 

This does not seem to work for regular data types like `int` and `str`

```py
>>> from pydantic import BaseModel
>>> class BitToBool(BaseModel):
  2     val: int
  3     class Config:
  4         json_encoders = {int: lambda i: bool(i)}

>>> bit_bool = BitToBool(val=1)

>>> bit_to_bool.__json_encoder__
Out[179]: functools.partial(<function custom_pydantic_encoder at 0x7f749c0e79e0>, 
{<class 'int'>: <function BitToBool.Config.<lambda> at 0x7f747c15d320>})

>>> bit_to_bool.dict()
Out[180]: {'val': 1}

>>> bit_to_bool.json()
Out[181]: '{"val": 1}'
...
```
Same thing happens for strings where it is the same story > database field goes into HTTP post with something prepended to satisfy requirements.

```py
>>> class PrependCustomField(BaseModel):
  2     value: str
  3     class Config:
  4         json_encoders = {str: lambda s: f"CF{s}"}

>>> cf = PrependCustomField(value="EventSummary")

>>> cf.json()
Out[189]: '{"value": "EventSummary"}'
```

This does work with a custom class, this is just extra complexity and another class that isn't really intended to use for any other purpose:

```py
>>> class BitBool:
  2     __slots__ = ["_bit"]
  3
  4     def __init__(self, bit: int = 0):
  5         self._bit = bit
  6
  7     def __bool__(self):
  8         if self._bit == 0:
  9             return False
 10         if self._bit == 1:
 11             return True

>>> class BitToBool(BaseModel):
  2     val: BitBool
  3     class Config:
  4         arbitrary_types_allowed = True
  5         json_encoders = {BitBool: lambda i: bool(i)}

>>> bit_to_bool = BitToBool(val=BitBool(1))
>>> bit_to_bool.json()
Out[208]: '{"val": true}'
```

_Note:_ I can get the right behavior with custom validators, but this isn't the best since really I might still want the original data via `dict()`. I really only need it to change during serialization. Is there something obvious I am missing here?

Thanks!

**Comment by layday** [comment]
Behind the scenes, Pydantic constructs a callback from your `json_encoders` which it passes to the `default` argument of `json.dumps`.  This is only called for types which the `json` module cannot serialise natively and `json` does not make it easy to override this behaviour.

**Comment by randallpittman** [comment]
I can't serialize PurePosixPath if it is in a contained model--not sure if this is related (Pydantic 1.8.1)

```python
from pathlib import PurePosixPath
from typing import List

from pydantic import BaseModel

class Child(BaseModel):
    p: PurePosixPath
    class Config:
        arbitrary_types_allowed = True
        json_encoders = {
            PurePosixPath: lambda p: str(p)
        }

class Parent(BaseModel):
    children: List[Child]

child = Child(p=PurePosixPath("/data/tmp"))
print(child.json())
```
returns `{"p": "/data/tmp"}`

```python
parent = Parent(children=[child])
print(parent.json())
````

returns `TypeError: Object of type 'PurePosixPath' is not JSON serializable`.

**Comment by SvenPVoigt** [comment]
This is also the case for any custom defined classes.

```python
from pydantic import BaseModel
        
class UserDefined:
    pass

class Test(BaseModel):
    name: str
    other: UserDefined=None
        
    class Config:
        arbitrary_types_allowed: True
        json_encoders = {
            UserDefined: lambda v: str(v.x)
        }

something = UserDefined()
something.x = 11

test = Test(other=something)
print(test.json(indent=2))
```

This example gives the error: RuntimeError: no validator found for <class '__main__.UserDefined'>, see `arbitrary_types_allowed` in Config

However, the json_encoder gives and encoding function for the UserDefined class.

I think I see the error in `pydantic/json.py` and would like to address this. What is a good way of contributing? I am not sure how cython works though and how pydantic incorporates it. Is there a way to install pydantic without compiling so that I can test changes to the python code without recompiling to c? If not, how would I recompile pydantic?

**Comment by layday** [comment]
You have

```py
        arbitrary_types_allowed: True
```

This should be:

```py
        arbitrary_types_allowed = True
```

**Comment by SvenPVoigt** [comment]
Thanks for the catch.

Any idea why changing UserDefined to

```python
class UserDefined(BaseModel):
    x: int=None
```
Causes it to serialize as json (as BaseModel does) rather than the custom defined encoding?

**Comment by layday** [comment]
Pydantic models are converted to dictionaries prior to being encoded.

**Comment by tstorek** [comment]
I have the same problem. I like to define a list of str in my model but when calling model.json() I'd like to concatenate the str as comma-separated strings. 

I think that this would be especially useful to organize query parameters. Just think of a request model in which I can simply add and remove queries from a list instead of processing a whole list. 

1. I managed to realize this behavior by writing my own intermediate class. But this seems kind of ugly.
2. On a  second try I wanted the intermediate class to inherit from string and add validators to it. I tried to use pydantic.networks as a blueprint. But Once I inherited from String I couldn't use the custom serialization anymore.

This issue already cost me a lot of time. I tried to understand how the custom_serialization is injected.
As I understand the solution would be to calls custum_serialization before calling builtin json.dumps()?

It would be nice if also List[MyOtherType] could be serialized in a custom way.

I there a way to understand this behavior a little better? PyCharm doesn't let me step through :/
